rest/src/main/scala/io/udash/rest/openapi/OpenApiMetadata.scala

```diff
@@ -166,27 +176,41 @@ case class OpenApiGetOperation[T](
   def requestBody(resolver: SchemaResolver): Opt[RefOr[RequestBody]] = Opt.Empty
 }
 
+case class OpenApiCustomBodyOperation[T](
+  name: String,
+  methodTag: HttpMethodTag,
+  operationAdjusters: List[OperationAdjuster],
+  pathAdjusters: List[PathItemAdjuster],
+  parameters: List[OpenApiParameter[_]],
+  @encoded @rpcParamMetadata @tagged[Body] singleBody: OpenApiBody[_],
+  resultType: RestResultType[T]
+) extends OpenApiOperation[T] {
+  def requestBody(resolver: SchemaResolver): Opt[RefOr[RequestBody]] =
+    singleBody.requestBody(resolver).opt
+}
+
 case class OpenApiBodyOperation[T](
   name: String,
   methodTag: HttpMethodTag,
   operationAdjusters: List[OperationAdjuster],
   pathAdjusters: List[PathItemAdjuster],
   parameters: List[OpenApiParameter[_]],
-  @multi @rpcParamMetadata @tagged[BodyField] bodyFields: List[OpenApiBodyField[_]],
-  @optional @encoded @rpcParamMetadata @tagged[Body] singleBody: Opt[OpenApiBody[_]],
-  @isAnnotated[FormBody] formBody: Boolean,
+  @multi @rpcParamMetadata @tagged[Body] bodyFields: List[OpenApiBodyField[_]],
+  @reifyAnnot bodyTypeTag: BodyTypeTag,
   resultType: RestResultType[T]
 ) extends OpenApiOperation[T] {
 
   def requestBody(resolver: SchemaResolver): Opt[RefOr[RequestBody]] =
-    singleBody.map(_.requestBody(resolver).opt).getOrElse {
-      if (bodyFields.isEmpty) Opt.Empty else Opt {
-        val fields = bodyFields.iterator.map(p => (p.info.name, p.schema(resolver))).toList
-        val requiredFields = bodyFields.collect { case p if !p.info.hasFallbackValue => p.info.name }
-        val schema = Schema(`type` = DataType.Object, properties = Mapping(fields), required = requiredFields)
-        val mimeType = if (formBody) HttpBody.FormType else HttpBody.JsonType
-        RefOr(RestRequestBody.simpleRequestBody(mimeType, RefOr(schema), requiredFields.nonEmpty))
+    if (bodyFields.isEmpty) Opt.Empty else Opt {
+      val fields = bodyFields.iterator.map(p => (p.info.name, p.schema(resolver))).toList
+      val requiredFields = bodyFields.collect { case p if !p.info.hasFallbackValue => p.info.name }
+      val schema = Schema(`type` = DataType.Object, properties = Mapping(fields), required = requiredFields)
+      val mimeType = bodyTypeTag match {
+        case _: JsonBody => HttpBody.JsonType
+        case _: FormBody => HttpBody.FormType
+        case _ => throw new IllegalArgumentException(s"Unexpected body type $bodyTypeTag")
```

Comment:
`BodyTypeTag` is sealed and only has a couple more implementations, you could explicitly check all cases here to enable exhaustivity verification for future changes.